src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionProvider.cs

```diff
@@ -261,31 +268,75 @@ internal override async Task<CompletionChange> GetChangeAsync(Document document,
+                return Utilities.Collapse(originalText, changes.ToImmutableArray());
+            }
+
+            async Task<(bool shouldFullyQualify, bool needsSimplification)> ShouldCompleteWithFullyQualifyTypeName()
             {
                 var workspace = document.Project.Solution.Workspace;
 
                 // Certain types of workspace don't support document change, e.g. DebuggerIntellisense
                 if (!workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
                 {
-                    return true;
+                    return (true, false);
                 }
 
                 // During an EnC session, adding import is not supported.
                 var encService = workspace.Services.GetService<IDebuggingWorkspaceService>()?.EditAndContinueServiceOpt;
                 if (encService?.EditSession != null)
                 {
-                    return true;
+                    return (true, false);
                 }
 
                 // Certain documents, e.g. Razor document, don't support adding imports
                 var documentSupportsFeatureService = workspace.Services.GetService<IDocumentSupportsFeatureService>();
                 if (!documentSupportsFeatureService.SupportsRefactorings(document))
                 {
-                    return true;
+                    return (true, false);
                 }
 
-                return false;
+                // We might need to qualify unimported types to use them in an import directive, because they only affect members of the containing
+                // import container (e.g. namespace/class/etc. declarations).
+                //
+                // For example, `List` and `StringBuilder` both need to be fully qualified below: 
+                // 
+                //      using CollectionOfStringBuilders = System.Collections.Generic.List<System.Text.StringBuilder>;
+                //
+                // However, if we are typing in an C# using directive that is inside a nested import container (i.e. inside a namespace declaration block), 
+                // then we can add an using in the outer import container instead (this is not allowed in VB). 
+                //
+                // For example:
+                //
+                //      using System.Collections.Generic;
+                //      using System.Text;
+                //
+                //      namespace Foo
+                //      {
+                //          using CollectionOfStringBuilders = List<StringBuilder>;
+                //      }
+                //
+                // Here we will always choose to qualify the unimported type, just to be consistent and keeps things simple.
+                var syntaxContext = await CreateContextAsync(document, completionListSpan.Start, cancellationToken).ConfigureAwait(false);
+                return (syntaxContext.IsInImportsDirective, true);
```

Comment:
this si not actually desirable.  users have indicated that they would prefer fully qualified names in imports so that they know what symbol this is actually resolving to.

Plus, you end up needing a massive amount of code to simplify this (which is also surpirsing) which i would prefer just be gone :)

Reply:
or, alternatively, why not just check:

1. if you're on the RHS of an alias, and you are *nested*, only then should you use the minimal name, and add the import.
2. otherwise, if you're on the RHS of an alias, just spit out hte fully qualified name.

Then there's no need to do any sort of post-simplification pass for this very minor case.

Reply:
OK, I prefer "always fully qualify", makes everything simpler. Will fix

Reply:
Done